### PR TITLE
Update `atan2` F32 tests to account for atan based implementations

### PR DESCRIPTION
**Pull request by zoddicus**
Used atan(y/x) as the basis for atan2 accuracy, since this will be less accurate than having a direct atan2 instruction.

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

**Review by ben-clayton — COMMENTED** [review]
LGTM, but due to the subtleties around all of this, I'd like final approval from @dneto0.

File: src/webgpu/util/f32_interval.ts [hunk 1]
```diff
@@ -746,26 +746,47 @@ const AtanIntervalOp: PointToIntervalOp = {
...
-          ulpInterval(kValue.f32.positive.pi.whole, numULP)
-        );
-      }
+      // For x > 0 atan2 is supposed to be 0, and for x < 0 atan2 is -π or π depending on the direction y/x approaches
+      // 0.
+      // Because implementations will be be calculating this via numeric approximations that will vary slightly around
+      // 0 and subnormals, it is very difficult to write a rule for accuracy here without just dictating a specific
+      // implementation. Instead if the value is at y/x = 0, then the result should be in [-π, π], plus 4096 ULP from
+      // atan.
+      return F32Interval.span(
```
Comment (dneto0) [inline src/webgpu/util/f32_interval.ts:768]: From the IEEE rules:

* If y is ±0 and x is negative or -0, ±π is returned
* If y is ±0 and x is positive or +0, ±0 is returned

Can we follow that here?   If  y === 0,  and X >0, then expect 0  within 4096 ULPs

Graphically, this follows the middle curve in this graph https://en.wikipedia.org/wiki/Atan2#/media/File:Arctangent2.svg to say that when x >0 then as y/x approaches zero, atan2(y,x) smoothly approaches 0.


Or do you already have evidence that drivers are erratic for  y == 0  and X> 0

File: src/webgpu/util/f32_interval.ts [hunk 2]
```diff
@@ -746,26 +746,47 @@ const AtanIntervalOp: PointToIntervalOp = {
...
+        ulpInterval(kValue.f32.negative.pi.whole, 4096),
+        ulpInterval(kValue.f32.positive.pi.whole, 4096)
+      );
     }
-    return ulpInterval(Math.atan2(y, x), numULP);
+
+    const atan_yx = atanInterval(divisionInterval(y, x));
```
Comment (dneto0) [inline src/webgpu/util/f32_interval.ts:774]: These parts look fine.

**Review by dneto0 — CHANGES_REQUESTED** [review]

File: src/webgpu/util/f32_interval.ts [hunk 3]
```diff
@@ -746,26 +746,47 @@ const AtanIntervalOp: PointToIntervalOp = {
...
+export function atanInterval(n: number | F32Interval): F32Interval {
   return runPointOp(toF32Interval(n), AtanIntervalOp);
 }
 
 const Atan2IntervalOp: BinaryToIntervalOp = {
   impl: (y: number, x: number): F32Interval => {
-    const numULP = 4096;
+    if (x === 0) {
+      // atan2 is not defined meaningfully here
+      return F32Interval.any();
```
Comment (dneto0) [inline src/webgpu/util/f32_interval.ts:757]: DXC and cppreference both say that

x = 0   y <0 should yield  -pi/2
x = 0  y  > 0 should yield pi / 2

File: src/webgpu/util/f32_interval.ts (hunk 1 above)
Comment (zoddicus) [reply src/webgpu/util/f32_interval.ts:768]: @amaiorano I believe there were Windows drivers that were returning ±π in unexpected cases, but I don't recall the details.  Do you recall?

Those cases may be covered by `If y is ±0 and x is negative or -0, ±π is returned`, so I will implement what you have sugggested and we can see if it is breaking anywhere.

Comment (zoddicus) [reply src/webgpu/util/f32_interval.ts:768]: Specifically this is becoming an issue when y is subnormal. Probably something related to y = 0 causing y/x = 0 and that is hard coded or something in the driver.

For example:
    (-1.1754942106924411e-38 (0x807fffff subnormal), 12.734692573547363 (0x414bc14d))
        returned: 3.1415927410125732 (0x40490fdb)
        expected: [-4.81482486096809e-35,4.81482486096809e-35]

Comment (zoddicus) [reply src/webgpu/util/f32_interval.ts:768]: Hmmm.... it appears that π is being returned unexpected when y is a negative subnormal, so maybe they are just doing -0 wrong?

File: src/webgpu/util/f32_interval.ts (hunk 3 above)
Comment (zoddicus) [reply src/webgpu/util/f32_interval.ts:757]: I agree those specific implementations do that, but if we are supporting other implementations where atan2 ~ atan(y/x) then those others one will get an undefined value when x = 0, and that would be more permissive than pi/2 or -pi/2.

I should probably rewrite the comment to be clear that it is in the atan(y/x) case that this is undefined.

File: src/webgpu/util/f32_interval.ts (hunk 1 above)
Comment (dneto0) [reply src/webgpu/util/f32_interval.ts:768]: Ok, it's very clarifying that it's the "y is subnormal" case.  The function is wildly approaching discontinuity from -pi to +pi as y transitions from small negative to small positive.
(Stare at the colour diagram at https://en.cppreference.com/w/cpp/numeric/math/atan2 )

Allowing for small errors in calculation it's not particularly meaningful to ask about the quadrant of a point near the origin.

Suggest we update the spec to say the accuracy of atan2 at most when y is normal.

**Comment by zoddicus** [comment]
This needs to be updated based on https://github.com/gpuweb/gpuweb/pull/3474

**Comment by zoddicus** [comment]
I have updated this PR to match the rules from https://github.com/gpuweb/gpuweb/pull/3474

**Review by dneto0 — APPROVED** [review]
Thanks for your patience.

I filed a Vulkan CTS issue internally about the fact that the y-is-subnormal case having out-of-spec results.  I pointed out your evidence for the conformant NVIDIA driver.
Vulkan CTS issue: #4022 (internal to Khronos).

Thanks!

**Pull request merged**